File: language/java_study/javaLearn/src/main/java/com/iamhotcenplotz/chapter25/pool/druid_/Demo01_.java
package com.iamhotcenplotz.chapter25.pool.druid_;

import com.alibaba.druid.pool.DruidDataSourceFactory;

import javax.sql.DataSource;
import java.io.FileInputStream;
import java.sql.Connection;
import java.util.Properties;

/**
 * @author: Der Hotcenplotz
 * @version： 1.0
 * @date: 2025-04-28 16:47
 * @description: 德鲁伊连接池
 */


public class Demo01_ {

    public static void main(String[] args) throws Exception {
        String conf = "src/main/java/com/iamhotcenplotz/chapter25/pool/druid_/druid.properties";
        Properties properties = new Properties();
        properties.load(new FileInputStream(conf));

        // 创建连接池
        DataSource connections = DruidDataSourceFactory.createDataSource(properties);

        // 测试
        long start = System.currentTimeMillis();
        for (int i = 0; i < 500000; i++) {
            Connection connection = connections.getConnection();
            connection.setAutoCommit(false);
            connection.close();
        }
        long end = System.currentTimeMillis();
        System.out.println("耗时：" + (end - start) + "ms");


    }
}

File: language/scala_study/ScalaLearn/src/main/scala/com/hotcenplotz/chapter08/self_/Demo01.scala
package com.hotcenplotz.chapter08.self_

/**
 * @author: Der Hotcenplotz
 * @version： 1.0 
 * @date: 2025-05-14 06:44
 * @description: 特质的自身类型示例代码
 * 特质的自身类型主要是为了解决特质的循环依赖问题。
 * */

object Demo01 {
    
}

// Logger就是自身类型特质，当做了自身类型后，
// 那么 trait Logger extends Exception，要求混入该特质的类也是Exception子类
trait Logger {
    // 明确告诉编译器，我就是Exception，如果没有这句话，下面的getMessage不能调用
    this: Exception =>
    
    def log():Unit = {
        // 既然我就是Exception，那么就可以调用其中的方法
        println(getMessage)
    }
}

// 不对！！！
// class Console extends Logger{}

// 对。
class Consoles extends Exception with Logger